npm/pkg/dataplane/ipsets/ipsetmanager.go

```diff
@@ -227,40 +250,100 @@ func (iMgr *IPSetManager) RemoveFromSets(removeFromSets []*IPSetMetadata, ip, po
 }
 
 func (iMgr *IPSetManager) AddToLists(listMetadatas, setMetadatas []*IPSetMetadata) error {
+	if len(listMetadatas) == 0 || len(setMetadatas) == 0 {
+		return nil
+	}
 	iMgr.Lock()
 	defer iMgr.Unlock()
 
-	if err := iMgr.checkForListMemberUpdateErrors(listMetadatas, setMetadatas, npmerrors.AppendIPSet); err != nil {
-		return err
+	// use this for efficient error checking later
+	memberNames := make(map[string]struct{}, len(setMetadatas))
+	// 1. check for errors in members and create any missing sets
+	for _, setMetadata := range setMetadatas {
+		setName := setMetadata.GetPrefixName()
+		set, exists := iMgr.setMap[setName]
+		if !exists {
+			// NOTE: any newly created IPSet will still be in the cache if an error is returned later
+			iMgr.createIPSet(setMetadata)
+			set = iMgr.setMap[setName]
+		}
+
+		// Nested IPSets are only supported for windows
+		// Check if we want to actually use that support
+		if set.Kind != HashSet {
+			return npmerrors.Errorf(npmerrors.AppendIPSet, false, fmt.Sprintf("ipset %s is not a hash set and nested list sets are not supported", setName))
+		}
+
+		memberNames[setName] = struct{}{}
 	}
 
 	for _, listMetadata := range listMetadatas {
+		// 2. create the list if it's missing and check for list errors
 		listName := listMetadata.GetPrefixName()
-		for _, setMetadata := range setMetadatas {
-			setName := setMetadata.GetPrefixName()
-			iMgr.addMemberIPSet(listName, setName)
+		list, exists := iMgr.setMap[listName]
+		if !exists {
+			// NOTE: any newly created IPSet will still be in the cache if an error is returned later
+			iMgr.createIPSet(listMetadata)
+			list = iMgr.setMap[listName]
+		}
+
+		if list.Kind != ListSet {
+			return npmerrors.Errorf(npmerrors.AppendIPSet, false, fmt.Sprintf("ipset %s is not a list set", listName))
+		}
+
+		if _, exists := memberNames[listName]; exists {
```

Comment:
I do not think we need this check if listname == membername since, in the above loop we are checking if the set is HashSet and in this loop we are checking if it is ListSet, And our PrefixName depends on the "kind" right ? so we guarantee the name to be unique ? i am okay with keeping this check for safety reasons, Wdyt ?

Reply:
makes sense. I'll replace with a comment